Mobile only: move interaction buttons into navbar

AllanYangZhou opened the pull request:
Currently mobile only. "Submit" and "Continue" buttons from interactions moved into the navbar.

For non-inline interactions, the "Continue" button remains in the help card rather than in the navbar because in some mobile devices the help card could cover up the bottom-right of the navbar.

Commit:
Mobile only: move interaction buttons into navbar

Commit:
Put back minified angular

seanlip commented on core/templates/dev/head/directives.js:
```diff
@@ -23,9 +23,17 @@ oppia.directive('angularHtmlBind', ['$compile', function($compile) {
   return {
     restrict: 'A',
     link: function(scope, elm, attrs) {
+      // Clean up old scopes if the html changes.
```
Interesting! Out of curiosity, why did you need this?

AllanYangZhou replied:
For example if the student enters an incorrect answer the interaction "refreshes" which changes the value of `angularHtmlBind`, which then re-compiles and creates a new identical-looking directive. However if you don't destroy the scope the old callback on 'progress-nav-submit' in the interaction directive's controller lingers around, so you now have two callbacks for that event when the student presses the submit button. And the old callback's answer is always processed first, so the student can never correct their answer.

seanlip replied:
You know, that sounds familiar -- at one point we had two calls to submitAnswer() coming in for each submit operation and we couldn't figure out why. I think you might have just solved this long-standing mystery :)

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -777,6 +777,10 @@ oppia.directive('conversationSkin', [
               $scope.explorationId);
           };
+
+          $scope.progressNavSubmit = function() {
```
$scope.submitAnswerFromProgressNav

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -777,6 +777,10 @@ oppia.directive('conversationSkin', [
           };
+
+          $scope.progressNavSubmit = function() {
+            $scope.$broadcast('progress-nav-submit');
```
Use a named constant similar to EVENT_ACTIVE_CARD_CHANGED, e.g. EVENT_ANSWER_SUBMITTED_FROM_NAV. Declare the constant in the same place as EVENT_ACTIVE_CARD_CHANGED and import it where needed.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -777,6 +777,10 @@ oppia.directive('conversationSkin', [
+
+          $scope.progressNavSubmit = function() {
+            $scope.$broadcast('progress-nav-submit');
+          }
```
Missing semicolon.

AllanYangZhou replied:
fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
       templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
         '/pages/exploration_player/' +
         'progress_nav_directive.html'),
+      scope: {
```
Excellent that we're introducing a scope here.

Could you move this key above templateUrl? I think we do that elsewhere.

AllanYangZhou replied:
fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
+              ExplorationPlayerStateService.isInteractionInline(
+                $scope.activeCard.stateName));
+            $scope.helpCardHasContinueButton = false;
+          }
```
Missing semicolon

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
+          ExplorationPlayerStateService, CONTINUE_BUTTON_FOCUS_LABEL) {
+          $scope.CONTINUE_BUTTON_FOCUS_LABEL = CONTINUE_BUTTON_FOCUS_LABEL;
+
+          var update = function() {
```
If you're not using this anywhere else, may be better to declare as an anonymous function inline on line 57. If you don't want to do this then this needs a clearer name (update what?)

AllanYangZhou replied:
changed to `updateActiveCardInfo`

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
+
+          var update = function() {
+            var index = PlayerPositionService.getActiveCardIndex();
+            $scope.activeCardIndex = index;
```
Why not just
```
    $scope.activeCardIndex = (
      PlayerPositionService.getActiveCardIndex());
```
(and drop the index var)?

AllanYangZhou replied:
Ah this was purely cosmetic because I use `index` later and it saves characters.

seanlip replied:
Ah ok. Mildly prefer what I suggested, for two reasons:
- Fewer different vars to keep in the reader's head
- "index" on its own isn't a very descriptive variable name.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
+            var index = PlayerPositionService.getActiveCardIndex();
+            $scope.activeCardIndex = index;
+            $scope.activeCard = PlayerTranscriptService.getCard(index);
+            $scope.transcriptLength = PlayerTranscriptService.getNumCards();
```
Throughout this method, don't declare stuff on $scope unless you really need to. E.g. I don't think you need to do that for transcriptLength.

Declaring stuff on $scope exposes it to the directive HTML and it would be better not to expose properties needlessly.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -23,20 +23,41 @@ oppia.directive('progressNav', [
-            $scope.transcriptLength = PlayerTranscriptService.getNumCards();
+          }, update);
+
+          $scope.$on('helpCardAvailable', function(event, helpCard) {
```
Generally we use evt, I think that's because event is a keyword in old versions of IE or something like that.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -47,13 +68,12 @@ oppia.directive('progressNav', [
-          $scope.hasNext = function() {
-            return $scope.currentCardIndex < $scope.transcriptLength - 1;
-          };
+          $scope.showContinueButton = function() {
```
Seems more like $scope.isContinueButtonShown (since it returns a boolean)

AllanYangZhou replied:
changed to `shouldContinueButtonBeShown`.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -47,13 +68,12 @@ oppia.directive('progressNav', [
+            var lastPair = $scope.activeCard.inputResponsePairs[
+              $scope.activeCard.inputResponsePairs.length - 1];
+            return $scope.isInteractionInline &&
+              $scope.activeCard.destStateName && lastPair.oppiaResponse;
```
Should this be lastPair.oppiaResponse !== null, or are you including an empty string as well?

Also, suggest casting to Boolean explicitly: `return Boolean(...);`. Otherwise I think it takes on the last value.

Also, just to check, is this a minimal criterion? I.e. are all three clauses necessary, and removing any one of them makes this check wrong?

AllanYangZhou replied:
Looks like it should be an undefined check, fixed. Though interestingly [this code](https://github.com/oppia/oppia/blob/4c0a9dfa559c7a183eb26cb12aec437ab6122ac8/core/templates/dev/head/pages/exploration_player/tutor_card_directive.html#L71) doesnt check against undefined?

Yes, all three should be true to show the continue button.

seanlip replied:
Ah, I didn't realize there was prior art. Let's keep it as it was then (no need for introducing !== undefined here). In the future we should get oppiaResponse to be a domain object and then this problem will go away. Sorry for the churn.

seanlip commented on core/templates/dev/head/pages/exploration_player/conversation_skin_directive.html:
```diff
@@ -47,7 +47,10 @@
       </div>
-      <progress-nav ng-if="isViewportNarrow()"></progress-nav>
+      <progress-nav ng-if="isViewportNarrow()"
+                    submit-fn="progressNavSubmit()"
```
on-submit

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/progress_nav_directive.html:
```diff
@@ -1,12 +1,40 @@
+        However, the current approach is cleaner in angular-1.5.8.
+      -->
+      <md-button ng-if="interaction.id === 'Continue'"
+        class="oppia-learner-confirm-button" ng-click="submitFn()">
```
Our convention in this codebase is to align the left edge of 'class' with the left edge of 'ng-if' in the previous line.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_player/progress_nav_directive.html:
```diff
@@ -1,12 +1,40 @@
+      </md-button>
+    </div>
+    <div ng-if="showContinueButton()">
+      <continue-button ng-if="!isIframed"
```
Optional: not sure you need the outer div. Why not just `<continue-button ng-if="showContinueButton() && !isIframed">`?

AllanYangZhou replied:
Yeah, I removed the outer div.

seanlip commented on core/templates/dev/head/pages/exploration_player/progress_nav_directive.html:
```diff
@@ -1,12 +1,40 @@
+        <[interaction.customizationArgs.buttonText.value]>
+        <i class="fa fa-arrow-right" style="font-size: 19px; padding-top: 1.5px;"></i>
+      </md-button>
+      <md-button ng-if="interaction.id === 'TextInput' ||
```
What about item selection input? Are there others?

I'm quite concerned about maintainability here. It would be better for each interaction's .py file to declare a includes_submit_button field instead of hard-coding here. Otherwise I don't see future interaction creators remembering to update this.

seanlip replied:
@AllanYangZhou and I discussed this in chat. Here's a summary:

- Let's add a boolean property show_generic_submit_button_in_nav to the interaction .py files. Set it to True for the six interactions mentioned here and False for the others. (Also check parity with other interactions, e.g. ItemSelectionInput.)
  - Add a comment above that property in base.py that's something like "Whether to show a submit button in the navigation area when the window width is small. Developers should ensure that their interaction JS binds the XXX event, and hide the submit button from the main interaction when the window width is small". 
  - Also, add a comment above that property in the Continue button saying "The Continue button does appear in the nav, but we handle it separately."
  - Then, here, do ng-if="interaction.show_submit_button".
- There's also another issue, which is that an interaction developer may, in the future, modify the code interaction's submit button to "Run and Submit", or add a play icon to it, or something.
  - Further work needs to be done so that a generic button is inserted on the bottom right of the interaction automatically, and there is no dev-written code for that button in particular in all of the existing interactions -- or, alternatively, there should be some backend/frontend test that ensures the conditions specified always hold (maybe pull the standard button into its own directive and check that the interaction includes that code snippet exactly). @AllanYangZhou will file an issue for this with more details.
  - In the meantime, let's add a comment above each Submit button in those interactions, and also above the corresponding JS code -- something like "this should be a standard Submit button otherwise it'll break parity with the interaction's show_generic_submit_button_in_nav setting".

AllanYangZhou replied:
Ah, one thing that didn't occur to me is that ItemSelectionInput is already an example of the 2nd bullet point, it shows a Submit button only in some cases (when maxAllowableSelectionCount > 1) .

So I'm not sure if we can put that off actually. EDIT: or we'd lose the dynamic behavior of the ItemSelectionInput's Submit on mobile.

seanlip replied:
Sigh. For now, let's special-case it like we do for the continue button. I can't think of a better way to do this that's still reasonable to implement in the short-term.

seanlip commented on core/templates/dev/head/pages/exploration_player/tutor_card_directive.html:
```diff
@@ -68,13 +68,17 @@
           </continue-button>
         </div>
-        <div ng-if="activeCard.interactionHtml && !activeCard.destStateName">
+        <!-- The seemingly redundant check for isInteractionInline is necessary
```
Great docstring!

seanlip commented on extensions/interactions/NumericInput/directives/NumericInput.js:
```diff
@@ -49,6 +51,14 @@ oppia.directive('oppiaInteractiveNumericInput', [
+            return WindowDimensionsService.isWindowNarrow();
+          };
+
+          $scope.$on('progress-nav-submit', function() {
```
Please file an issue to find a way to do this transparently -- i.e. we should have some common init() that is called for all interactions, and that init() should register the event handler if the interaction dict has an includes_submit_button attribute (see above).

seanlip commented:
I found two UI bugs when testing:
- When pressing the back button, it lingers for a while before disappearing. This is kind of confusing if I've just navigated back to the first card. The same thing happens when navigating forward to the last card.
- On screen sizes between 768 and 971 px (roughly), two submit buttons appear:

![bug](https://user-images.githubusercontent.com/10575562/34434700-ad0aff4a-ec3c-11e7-9f7d-d76938c4a1ae.png)

Commit:
Stop using isViewportNarrow for navbar

Commit:
Fix minor code issue

Commit:
More minor code fixed

Commit:
Resolve Merge Conflicts

Commit:
Add class attr to interactions to control whether progress nav shows submit button

Commit:
Include the other interactions too

seanlip commented on extensions/interactions/base.py:
```diff
@@ -111,6 +111,10 @@ class BaseInteraction(object):
+    # Whether to show a Submit button in the progress navigation area when the
+    # window width is narrow. The interaction JS must handle the
+    # EVENT_PROGRESS_NAV_SUBMITTED event broadcast by the nav Submit button.
+    show_nav_submit_button = False
```
Remember to update the backend tests that validate interactions.

AllanYangZhou replied:
Thanks for the tip!

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -31,25 +31,33 @@ oppia.directive('progressNav', [
                 $scope.activeCard.stateName));
+            $scope.interactionCustomizationArgs = interaction.customizationArgs;
+            $scope.interactionId = interaction.id;
+            $scope.showGenericSubmitButtonInNav = INTERACTION_SPECS[
```
$scope.genericSubmitButtonIsShown

(make it not start with a verb, and we're in the nav already so I think you can drop the nav part from the name.)

AllanYangZhou replied:
Got it

seanlip commented:
Hi @AllanYangZhou -- let me know when you'd like a full review, I just took a cursory glance because I wasn't sure if you're finished yet. Thanks!

Commit:
Fix backend test

Commit:
Fix bug with embed explorations

Commit:
Remove navigatino button lag

Commit:
Add comments about the Submit button

AllanYangZhou commented:
@seanlip PTAL, addressed comments and the UI bugs should also be fixed. Haven't filed the associated issues, will do that if I have time later tonight.

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -20,39 +20,69 @@ oppia.directive('progressNav', [
+          INTERACTION_SPECS) {
+          $scope.CONTINUE_BUTTON_FOCUS_LABEL = CONTINUE_BUTTON_FOCUS_LABEL;
+
+          var transcriptLength, isInteractionInline, interaction;
```
- Move `interaction` inside, since it's only needed within the scope of updateActiveCardInfo(). I.e. do "var interaction = ..." on line 51.
- Please rename isInteractionOutline to interactionIsInline so it doesn't start with a verb.
- Instantiate transcriptLength to 0, and interactionIsInline to true, on separate lines.

seanlip commented on extensions/interactions/base_test.py:
```diff
@@ -150,7 +150,8 @@ def test_interaction_properties(self):
             'narrow_instructions', 'needs_summary',
-            'default_outcome_heading', 'can_have_solution'])
+            'default_outcome_heading', 'can_have_solution',
+            'show_nav_submit_button'])
```
Still need to modify test_default_interactions_are_valid(), particularly the _INTERACTION_CONFIG_SCHEMA. Always try and find ways to ensure that the standard conventions you're setting up persist into the future.

seanlip commented on extensions/interactions/CodeRepl/directives/CodeRepl.js:
```diff
@@ -34,216 +34,231 @@ oppia.directive('oppiaInteractiveCodeRepl', [
+          $scope, $attrs, WindowDimensionsService,
+          UrlService, EVENT_PROGRESS_NAV_SUBMITTED) {
+          $scope.isSubmitHidden = function() {
+            return (!UrlService.isIframed() &&
```
Here and below: if a bracketed phrase spans multiple lines, break after the first '(':

```
            return (
              !UrlService.isIframed() &&
              WindowDimensionsService.isWindowNarrow());
```

seanlip commented on extensions/interactions/Continue/directives/Continue.js:
```diff
@@ -30,30 +30,42 @@ oppia.directive('oppiaInteractiveContinue', [
+      controller: [
+        '$scope', '$attrs', 'WindowDimensionsService',
+        'UrlService', 'EVENT_PROGRESS_NAV_SUBMITTED',
+        function($scope, $attrs, WindowDimensionsService,
```
Same thing here and elsewhere:

```
        function(
            $scope, $attrs, WindowDimensionsService,
            UrlService, EVENT_PROGRESS_NAV_SUBMITTED) {
          $scope.buttonText...
```

The extra indent of 2 for the function args sets them apart from the line after that where the inner scope actually starts.

seanlip reviewed (changes requested):
Some minor code changes left, but looks good!

I tested the UI. There's a bug:
- Load the "all interactions" demo exploration from the admin page, then go to http://localhost:8181/explore/16.
- Play to the fourth card (image region). Make the window small. Click on various blocks until you're correct. A card appears at the bottom right.
- Don't click on that card; click on the Back button instead.
- Click forward again. The card's disappeared. There is no way to continue and the student is permanently stuck.

Commit:
Style issues, extend backend test

Commit:
Condense interaction submit i18ns

Commit:
Add isAnswerValid flag

seanlip commented on core/templates/dev/head/pages/exploration_player/ProgressNavDirective.js:
```diff
@@ -20,39 +20,72 @@ oppia.directive('progressNav', [
               PlayerPositionService.setActiveCardIndex(index);
               $rootScope.$broadcast('updateActiveStateIfInEditor',
                 PlayerPositionService.getCurrentStateName());
             }
```
perhaps `else throw Error` here -- the alternative can't happen, right?

AllanYangZhou replied:
Done

seanlip commented on core/templates/dev/head/services/ExplorationHtmlFormatterService.js:
```diff
@@ -52,6 +52,12 @@ oppia.factory('ExplorationHtmlFormatterService', [
         if (labelForFocusTarget) {
           element.attr('label-for-focus-target', labelForFocusTarget);
         }
+        // We cannot restrict this attr to interactions with
```
Provide more detail about what this means, and the possible values of answerValidity (bool?). Specify whether the parent need to have a setInteractionAnswerValidity() function, or if not having one is optional; specify what happens when no setInteractionAnswerValidity() function is defined on the parent.

AllanYangZhou replied:
Got it

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -777,6 +782,15 @@ oppia.directive('conversationSkin', [
+          $scope.interactionAnswerIsValid = true;
+          $scope.setInteractionAnswerValidity = function(answerValidity) {
+            $scope.interactionAnswerIsValid = answerValidity;
+          }
```
Missing semicolon.

Perhaps add a note that this is needed here in order to be available for the submit button in the navbar? I think we should be quite liberal in documenting the reasoning behind all this, since it's introducing temporary tangle to the codebase.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/TutorCardDirective.js:
```diff
@@ -179,6 +184,12 @@ oppia.directive('tutorCard', [
+            $scope.onChangeInteractionAnswerValidity({
+              answerValidity: answerValidity
+            });
+          }
```
Missing semicolon.

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/services/ExplorationHtmlFormatterService.js:
```diff
@@ -52,6 +52,12 @@ oppia.factory('ExplorationHtmlFormatterService', [
         }
+        // We cannot restrict this attr to interactions with
+        // show_nav_submit_button because of special cases such
+        // as ItemSelectionInput.
```
Also, let's specify this new attr only for interactions that have show_submit_button_in_nav, plus ItemSelectionInput. We should also put a comment in both places saying that these two things should match up with each other.

In addition, I think we should specify the setAnswerValidity function in the isolate scope for all interactions in which that functionality is provided. It should become part of the standard infrastructure for those interactions. To validate that what you've done generalizes beyond just ItemSelection, you might want to implement disabling of the submit button for either number/text -- this should be straightforward to do, and it provides assurance that what you have here works in general. We may also want to add a comment above that scope attr to say something like "Devs should ensure that this is called every time the answer is changed, even prior to submission." -- since otherwise I can see them forgetting to do it.

AllanYangZhou replied:
Sure! For Something like TextInput, what would the criteria for valid be, non-empty? Isn't it technically possible that the answer is an empty string?

seanlip replied:
Yeah, let's go with non-empty. In all the years, I've never seen an empty string being a valid answer.

AllanYangZhou replied:
Got it.

seanlip reviewed (commented):
Thanks @AllanYangZhou -- took a quick pass, looks good! A few remaining comments but they should be straightforward.

Commit:
Address review

seanlip commented on core/templates/dev/head/services/ExplorationHtmlFormatterService.js:
```diff
@@ -52,12 +55,20 @@ oppia.factory('ExplorationHtmlFormatterService', [
+        // case which has a special interaction-specific Submit
+        // button, not covered by show_nav_submit_button.
+        var navSubmitButtonExists = (
+          INTERACTION_SPECS[ interactionId].show_nav_submit_button ||
```
nit: remove space after [

AllanYangZhou replied:
Done.

seanlip commented on extensions/interactions/TextInput/directives/text_input_interaction_directive.html:
```diff
@@ -7,6 +7,7 @@
     <!-- This ng-click is needed for specifying solutions in the solution editor -->
     <md-button ng-if="!isSubmitHidden()" class="oppia-learner-confirm-button"
                ng-click="submitAnswer(answer)" style="margin-top: 15px;"
+               ng-disabled="!(answer.length > 0)"
```
answer.length === 0 is simpler?

AllanYangZhou replied:
Done.

seanlip commented on extensions/interactions/CodeRepl/directives/CodeRepl.js:
```diff
@@ -29,7 +29,8 @@ oppia.directive('oppiaInteractiveCodeRepl', [
         onSubmit: '&',
-        getLastAnswer: '&lastAnswer'
+        getLastAnswer: '&lastAnswer',
+        setAnswerValidity: '&' // Must call whenever the answer changes.
```
Couple of nits (sorry):
- We normally always have comments on top of the line they refer to.
- Can we say `This should be called whenever the answer changes.`? I think the current formulation is slightly unclear; my first thought was "what does this thing need to call?"

AllanYangZhou replied:
Got it.

seanlip reviewed (approved):
Remaining comments are nits, so giving LGTM. (Also, as discussed, please feel free to do desktop in this PR if it's easy, or leave it for a second one if it's not.) Thanks @AllanYangZhou!

Commit:
Address nits

seanlip commented on extensions/interactions/CodeRepl/directives/CodeRepl.js:
```diff
@@ -30,7 +30,8 @@ oppia.directive('oppiaInteractiveCodeRepl', [
         onSubmit: '&',
         getLastAnswer: '&lastAnswer',
-        setAnswerValidity: '&' // Must call whenever the answer changes.
+        // Must call whenever the answer changes.
```
Forgot to change this one :)

AllanYangZhou replied:
Oops, fixed.

seanlip commented on extensions/interactions/FractionInput/directives/FractionInput.js:
```diff
@@ -19,11 +19,12 @@ oppia.directive('oppiaInteractiveFractionInput', [
       templateUrl: UrlInterpolationService.getExtensionResourceUrl(
-        '/interactions/FractionInput/directives/' +
-        'fraction_input_interaction_directive.html'),
+      '/interactions/FractionInput/directives/' +
```
The previous indentation should be preserved?

AllanYangZhou replied:
Ah thanks for catching that, fixed.

Commit:
Fix
